NDTensors/src/lib/Sectors/src/category_product.jl

```diff
@@ -38,23 +52,89 @@ category_show(io::IO, k, v) = print(io, v)
 
+function Base.isless(s1::C, s2::C) where {C<:CategoryProduct}
+  return isless(
+    category_label.(values(categories(s1))), category_label.(values(categories(s2)))
+  )
+end
+
+# ==============  Cartesian product  =================
 ×(c1::AbstractCategory, c2::AbstractCategory) = ×(CategoryProduct(c1), CategoryProduct(c2))
 function ×(p1::CategoryProduct, p2::CategoryProduct)
   return CategoryProduct(categories_product(categories(p1), categories(p2)))
 end
 
-categories_product(l1::NamedTuple, l2::NamedTuple) = union_keys(l1, l2)
+function categories_product(l1::NamedTuple, l2::NamedTuple)
+  if length(intersect_keys(l1, l2)) > 0
+    throw(error("Cannot define product of shared keys"))
+  end
+  return union_keys(l1, l2)
+end
+
+# edge cases
+categories_product(l1::NamedTuple, l2::Tuple{}) = l1
+categories_product(l1::Tuple{}, l2::NamedTuple) = l2
 
 categories_product(l1::Tuple, l2::Tuple) = (l1..., l2...)
 
+×(a, g::AbstractUnitRange) = ×(to_graded_axis(a), g)
+×(g::AbstractUnitRange, b) = ×(g, to_graded_axis(b))
 ×(nt1::NamedTuple, nt2::NamedTuple) = ×(CategoryProduct(nt1), CategoryProduct(nt2))
 ×(c1::NamedTuple, c2::AbstractCategory) = ×(CategoryProduct(c1), CategoryProduct(c2))
 ×(c1::AbstractCategory, c2::NamedTuple) = ×(CategoryProduct(c1), CategoryProduct(c2))
 
-#
-# Dictionary-like implementation
-#
+function ×(l1::LabelledNumbers.LabelledInteger, l2::LabelledNumbers.LabelledInteger)
+  c3 = LabelledNumbers.label(l1) × LabelledNumbers.label(l2)
+  m3 = LabelledNumbers.unlabel(l1) * LabelledNumbers.unlabel(l2)
+  return LabelledNumbers.LabelledInteger(m3, c3)
+end
+
+function ×(g1::AbstractUnitRange, g2::AbstractUnitRange)
+  # keep F convention in loop order
+  v = [
+    l1 × l2 for l2 in BlockArrays.blocklengths(g2) for l1 in BlockArrays.blocklengths(g1)
+  ]
+  return GradedAxes.gradedrange(v)
+end
+
+# ===================  Fusion rules  ====================
+# generic case: fusion returns a GradedAxes, even for fusion with Empty
+function fusion_rule(::SymmetryStyle, s1::CategoryProduct, s2::CategoryProduct)
+  return to_graded_axis(categories_fusion_rule(categories(s1), categories(s2)))
+end
 
+# Abelian case: fusion returns CategoryProduct
+function fusion_rule(::AbelianGroup, s1::CategoryProduct, s2::CategoryProduct)
+  return categories_fusion_rule(categories(s1), categories(s2))
+end
+
+# Empty case: use × cast conventions between NamedTupled and ordered implem
+function fusion_rule(::EmptyCategory, s1::CategoryProduct, s2::CategoryProduct)
+  return s1 × s2
+end
+
+# ==============  Ordered implementation  =================
+CategoryProduct(t::Tuple) = _CategoryProduct(t)
+CategoryProduct(cats::AbstractCategory...) = CategoryProduct((cats...,))
+
+categories_equal(o1::Tuple, o2::Tuple) = (o1 == o2)
+
+sector(args...; kws...) = CategoryProduct(args...; kws...)
+
+function trivial(::Type{<:CategoryProduct{T}}) where {T<:Tuple}
+  return sector(ntuple(i -> trivial(fieldtype(T, i)), fieldcount(T)))
+end
```

Comment:
Looks like a good use case for defining an inner function and using some helper functions, i.e.:
```julia
categories_type(::Type{<:CategoryProduct{T}) where {T} = T

function trivial(type::Type{<:CategoryProduct})
  return categories_trivial(categories_type(type))
end

function categories_trivial(type::Type{<:Tuple})
  # Implementation
end

function categories_trivial(type::Type{<:NamedTuple})
  # Implementation
end
```